src/serialrw.rs

```diff
@@ -84,76 +96,202 @@ impl<R: AsyncRead + Unpin + Send> SerialFrameReader<R> {
 }
 #[async_trait]
 impl<R: AsyncRead + Unpin + Send> FrameReader for SerialFrameReader<R> {
-    async fn receive_frame(&mut self) -> crate::Result<RpcFrame> {
-        let mut has_stx = false;
+    async fn receive_frame_meta_data(&mut self) -> crate::Result<RpcFrameReception> {
+        Ok(
+        if self.meta.is_none() {
+            RpcFrameReception::Meta(self.receive_frame_meta().await?)
+        } else {
+            RpcFrameReception::Frame(self.receive_frame_data().await?)
+        })
+    }
+
+    async fn receive_frame_meta(&mut self) -> crate::Result<MetaMap> {
+        self.meta = None;
+        self.has_stx = false;
         'read_frame: loop {
-            if !has_stx {
+            if !self.has_stx {
                 loop {
                     match self.get_escaped_byte().await? {
                         Byte::Stx => { break }
                         _ => { continue }
                     }
                 }
             }
-            has_stx = false;
-            let mut data: Vec<u8> = vec![];
+            self.has_stx = false;
+            self.data = vec![];
             loop {
                 match self.get_escaped_byte().await? {
                     Byte::Stx => {
-                        has_stx = true;
+                        self.has_stx = true;
                         continue 'read_frame
                     }
-                    Byte::Data(b) => { data.push(b) }
-                    Byte::Etx => { break }
-                    _ => { continue 'read_frame }
+                    Byte::Data(b) => self.data.push(b),
+                    Byte::Etx => break,
+                    _ => continue 'read_frame,
+                }
+                if self.data.len() > 1 {
+                    let protocol = self.data[0];
+                    if protocol != Protocol::ChainPack as u8 {
+                        log!(target: "Serial", Level::Debug, "Not chainpack message");
+                        continue 'read_frame
+                    }
+                    let mut buffrd = BufReader::new(&self.data[1 ..]);
+                    let mut rd = ChainPackReader::new(&mut buffrd);
+                    if let Ok(Some(meta)) = rd.try_read_meta() {
```

Comment:
It seems that try_read_meta() is called on every single character received until valid meta data is parsed

Reply:
Yes, that's true, not very optimal way, but there is no control character between metadata and data. Any suggestions to do it better?